### ISSUE TITLE
Manage selected state in the data-table component instead of in the data itself

### ISSUE THREAD

**Issue by jakobrosenberg** [vuetifyjs/vuetify#400]
Currently the data-table stores its state within the data itself, which I find highly problematic.

**Storing view state in data models**

**Pros**
Not sure

**Cons**
- You have to map selected state to every data item.
- That gets complicated if your data is reactive.
- And even more complicated if your data is immutable, e.g. from a data store like Vuex.
- And even more complicated if the same data is used in a table more than once as the selected state is remembered in the store.
- And even more complicated if your server enforces schema or doesn't sanitize your submitted data
- Lastly, you risk data corruption in case your data entries already have a value property and you're unaware that the selected state by default overwrites the value.

Many of the cons listed can be tackled through various boilerplate, methods, computed methods, watchers, server sanitation etc., but the crux of the matter is, view state shouldn't leak into the data in the first place.

The current data-table uses the following scope 
```
  props: {
    item: {
      value: false,
      name: 'Ice cream sandwich',
      calories: 237,
      fat: 9.0,
      carbs: 37,
      protein: 4.3,
      sodium: 129,
      calcium: '8%',
      iron: '1%'
    }
  }
```

I propose this
```
  props: {
    checked: false,
    item: {
      name: 'Ice cream sandwich',
      calories: 237,
      fat: 9.0,
      carbs: 37,
      protein: 4.3,
      sodium: 129,
      calcium: '8%',
      iron: '1%'
    }
  }
```

Move the checked state out of the item and into props and keep the logic inside the data table component. The selected state can be be queried through $ref as well as emitted from the data table component.

**Comment by johnleider** [comment]
Could you provide some pseudo code in what a data table example would look like with this implementation?

**Comment by jakobrosenberg** [comment]
The data table example would remain the same as that in the current documentation except for the omission of ``value: false`` in each row.

```
{
   value: false, //omitted
   name: 'Ice cream sandwich',
   calories: 237,
   fat: 9.0,
   carbs: 37,
   protein: 4.3,
   sodium: 129,
   calcium: '8%',
   iron: '1%'
 }
```

An additional ``key`` property, like the one used in ``v-for`` could prove helpful.


The current method of getting checked items looks like this
``checkedItems = this.myItems.filter(row => row.value)``

With a ref it would look like this
``checkedItems = this.$ref.myTable.rows.filter(row => row.checked)``

Additionally you can emit the selected the selected rows to the parent component. I'm not sure which would be easier or if there'd be reactivity issues.


All in all, there should be very little change to the documentation as all we're doing is moving data-tables checked state to the data-table itself.

**Comment by jakobrosenberg** [comment]
Considering the amount of data passed to and from a data table, it would be beneficial to create a model for table data.

With the new [child/parent data syncing](https://github.com/vuejs/vue/issues/4946#issuecomment-292065374) that's planned for Vue, the following would be example.

**Basic example** (table model is optional)

```javascript
export default {
  data() {
    return {
      items: myStore.items,
    }
  }
}
```

```html
<v-data-table v-model="items">
```


**with table model**
```javascript
export default {
  data() {
    return {
      items: myStore.items,
      tableModel1: {
        //optional properties. If not defined, v-data-table will use default values
        itemsPerPage: 10,
        currentPage: 1
      },
    }
  }
}
```

```
<v-data-table v-model="items" :table-model="tableModel1" >
```

**Async example**
```javascript
export default {
  data() {
    return {
      tableModel2: {}
    }
  },
  computed: {
    async computedItems() {      
      const {currentPage, itemsPerPage} = this.tableModel2
      
      const results = await queryMyHost(
        offset: (currentPage-1 * itemsPerPage),
        limit: itemsPerPage
      )
      return results
    }
  }
}
```



```html
<v-data-table v-model="computedItems" :table-model="tableModel2" >
```


**Programatically change page example**
```javascript
export default {
  data() {
    return {
      items: myStore.items,
      tableModel1: {
        //optional properties. If not defined, just use default values
        itemsPerPage: 10,
        currentPage: 1
      },
      tableModel2: {}
    }
  },
  methods: {
    nextPage() {
      this.tableModel1.currentPage++
    }
  }
}
```

```html
<v-data-table v-model="items" :table-model="tableModel1" >
```

By providing the simple table model, it would be possible to use the table properties in watchers and computed data, as well as programatically update table properties. All the above examples require no boilerplate beyond their own logic.

**Comment by jakobrosenberg** [comment]
Additionally the table model would make it possible to manually insert table headers and footers as such

```html
<v-card>
  <v-card-row>
    <v-data-table v-model="items" :table-model=""tableModel1">
  </v-card-row>
  <v-card-row>
    <v-btn primary>Add item</v-btn>
    <v-spacer>
    <v-data-table-pagination :table-model="tableModel1">
  </v-card-row>
</v-card>

```

**Comment by nekosaur** [comment]
Implemented in #516
